qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,139 @@
+        """
+        Get the expected numa cpus for every node according to the qemu cli
+        """
+
+        def _get_boundary(value, max_value, weight):
+            min_boundary = int(value if value else 0) * weight
+            max_boundary = int(value if value else (max_value - 1)) * weight
+            return (min_boundary, max_boundary)
+
+        error_context.context("Get the expected cpus", logging.info)
+        cpus = params.objects("guest_numa_cpus")
+        expected_cpus = []
+        tmp = {}
+        for cpu in cpus:
+            nodeid = params["numa_cpu_nodeid_%s" % cpu]
+            socket = params.get("numa_cpu_socketid_%s" % cpu)
+            die = params.get("numa_cpu_dieid_%s" % cpu)
+            core = params.get("numa_cpu_coreid_%s" % cpu)
+            thread = params.get("numa_cpu_threadid_%s" % cpu)
+            if vm_arch in ('x86_64', 'i386'):
+                socket_min, socket_max = _get_boundary(socket, vcpu_sockets, socket_weight)
+                die_min, die_max = _get_boundary(die, vcpu_dies, die_weight)
+                core_min, core_max = _get_boundary(core, vcpu_cores, core_weight)
+                thread_min, thread_max = _get_boundary(thread, vcpu_threads, thread_weight)
+                cpu_min = socket_min + die_min + core_min + thread_min
+                cpu_max = socket_max + die_max + core_max + thread_max
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += range(cpu_min, cpu_max+1, 1)
+                else:
+                    tmp[nodeid] = range(cpu_min, cpu_max+1, 1)
+            elif vm_arch in ('ppc64le'):
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += [int(core)]
+                else:
+                    tmp[nodeid] = [int(core)]
+        for item in sorted(tmp.items(), key=lambda item: item[0]):
+            expected_cpus.append(set(sorted(item[1])))
+        return expected_cpus
+
+    def get_hotpluggable_cpus():
+        """
+        Parse output of "query-hotpluggable-cpus", get the cpus that bind to
+        special node.
+        """
+        error_context.context("Get the hotpluggable cpus", logging.info)
+        hotpluggable_cpus = []
+        tmp = {}
+        out = vm.monitor.info("hotpluggable-cpus")
+        for vcpu_info in out:
+            vcpu_info = vcpu_info["props"]
+            nodeid = vcpu_info.get("node-id")
+            if nodeid is not None:
+                socketid = vcpu_info.get("socket-id")
+                dieid = vcpu_info.get("die-id")
+                coreid = vcpu_info.get("core-id")
+                threadid = vcpu_info.get("thread-id")
+                if vm_arch in ('x86_64', 'i386'):
+                    cpu_index = socketid * socket_weight + dieid * die_weight \
+                                + coreid * core_weight + threadid * thread_weight
+                elif vm_arch in ('ppc64le'):
+                    # ppc only support core-id
+                    cpu_index = coreid
```

Comment:
> Should be `coreid * vcpu_threads`.

Not exactly, here the coreid is the value in "-numa cpu, coreid*",  it was defined by 'numa_cpu_coreid_cpu*'.

Since ppc only support coreid, the max value of coreid is `vcpu_socket * vcpu_coreid * vcpu_threads`.

Reply:
I wrote it wrongly. On ppc, the CPU index is `coreid // vcpu_threads`
```
# /usr/libexec/qemu-kvm -smp 8,threads=2 -vnc none -S -numa node,nodeid=0 -numa cpu,node-id=0,core-id=0 -qmp stdio
{"QMP": {"version": {"qemu": {"micro": 0, "minor": 1, "major": 5}, "package": "qemu-kvm-5.1.0-4.module+el8.3.0+7846+ae9b566f"}, "capabilities": ["oob"]}}
qemu-kvm: warning: Default splitting of RAM between nodes is deprecated, Use '-numa node,memdev' to explictly define RAM allocation per node
qemu-kvm: warning: CPU(s) not present in any NUMA nodes: CPU 1 [core-id: 2], CPU 2 [core-id: 4], CPU 3 [core-id: 6]
qemu-kvm: warning: All CPU(s) up to maxcpus should be described in NUMA config, ability to start up with partial NUMA mappings is obsoleted and will be removed in future
{"execute": "qmp_capabilities"}
{"return": {}}
{"execute": "query-cpus-fast"}
{"return": [{"arch": "ppc", "thread-id": 262820, "props": {"core-id": 0, "node-id": 0}, "qom-path": "/machine/unattached/device[1]/thread[0]", "cpu-index": 0, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262821, "props": {"core-id": 0, "node-id": 0}, "qom-path": "/machine/unattached/device[1]/thread[1]", "cpu-index": 1, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262822, "props": {"core-id": 2, "node-id": 0}, "qom-path": "/machine/unattached/device[2]/thread[0]", "cpu-index": 2, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262823, "props": {"core-id": 2, "node-id": 0}, "qom-path": "/machine/unattached/device[2]/thread[1]", "cpu-index": 3, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262824, "props": {"core-id": 4, "node-id": 0}, "qom-path": "/machine/unattached/device[3]/thread[0]", "cpu-index": 4, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262825, "props": {"core-id": 4, "node-id": 0}, "qom-path": "/machine/unattached/device[3]/thread[1]", "cpu-index": 5, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262826, "props": {"core-id": 6, "node-id": 0}, "qom-path": "/machine/unattached/device[4]/thread[0]", "cpu-index": 6, "target": "ppc64"}, {"arch": "ppc", "thread-id": 262827, "props": {"core-id": 6, "node-id": 0}, "qom-path": "/machine/unattached/device[4]/thread[1]", "cpu-index": 7, "target": "ppc64"}]}
```

You are using `vcpu_threads = 1` in this scenario, so you can see they are the same. But we should keep `cpu_index` matches to qemu.

Reply:
Oh~   yes, you are right, i tested with one power system:
1. To ensure all the cpus are bind to specified numa node, the number of `-numa cpu"` is "sockets * cores"
2. The `core-id` in `-numa cpu` is a multiple of 'threads'.
3. 'core-id' means `cpu_index`: range(core-id*threads, (core-id+1)*threads)

Is it right ?

Reply:
If check the output of `query-hotpluggable-cpus` with multiple threads, seems the cpu_index depends on the `vcpus-count`:

cli:
```
-object memory-backend-ram,size=1024M,id=mem-mem0 \
-object memory-backend-ram,size=1024M,id=mem-mem1 \
-object memory-backend-ram,size=1024M,id=mem-mem2 \
-object memory-backend-ram,size=1024M,id=mem-mem3  \
-smp 8,sockets=2,cores=2,threads=2 \
-numa node,memdev=mem-mem0,nodeid=0  \
-numa node,memdev=mem-mem1,nodeid=1  \
-numa node,memdev=mem-mem2,nodeid=2  \
-numa node,memdev=mem-mem3,nodeid=3  \
-numa cpu,node-id=0,core-id=0  \
-numa cpu,node-id=1,core-id=2  \
-numa cpu,node-id=2,core-id=4  \
-numa cpu,node-id=3,core-id=6  \
```
```
{"execute":"query-hotpluggable-cpus"}
{"return": [{"props": {"core-id": 6, "node-id": 3}, "vcpus-count": 2, "qom-path": "/machine/unattached/device[4]", "type": "host-spapr-cpu-core"}, {"props": {"core-id": 4, "node-id": 2}, "vcpus-count": 2, "qom-path": "/machine/unattached/device[3]", "type": "host-spapr-cpu-core"}, {"props": {"core-id": 2, "node-id": 1}, "vcpus-count": 2, "qom-path": "/machine/unattached/device[2]", "type": "host-spapr-cpu-core"}, {"props": {"core-id": 0, "node-id": 0}, "vcpus-count": 2, "qom-path": "/machine/unattached/device[1]", "type": "host-spapr-cpu-core"}]}
```

Take node-id 3 as an example, the core-id is 6, means the `cpu_index` is `range(core-id, core-id+vcpus-count)`, is it ?

Reply:
@yanan-fu Yes, `vcpu-count` of other architectures are 1, but on PPC, it's `threads`.
I think I may have a wrong understanding of the cpu_index you mentioned, My point is it's the index of hotpluggable unit (query-hotpluggable-cpus), like `CPU 0 [core-id: 0], CPU 1 [core-id: 2], CPU 2 [core-id: 4], CPU 3 [core-id: 6]`, so the index list is [0/2, 2/2, 4/2, 6/2]
But as you mentioned, it's used to `numa_cpu_coreid_cpu`, the index(core index) should be [0, 2, 4, 6].

Reply:
@PaulYuuu  Oh~, yes, we have different understanding of cpu_index here.  The cpu_index is the cpus for everynoed under `/sys/devices/system/node/node*/cpulist`.
Take the qemu cli in the comment above as an example:
node0 -> cpu0 -> coreid=0:  cpulist = 0-1
node1 -> cpu1 -> coreid=2:  cpulist = 2-3
node2 -> cpu2 -> coreid=4:  cpulist = 4-5
node3 -> cpu3 -> coreid=6:  cpulist = 6-7

Reply:
@PaulYuuu  Updated, consider multiple threads for ppc, could you help review again, thanks ~